playwright/_impl/_wait_helper.py

```diff
@@ -121,10 +126,37 @@ def listener(event_data: Any = None) -> None:
     def result(self) -> asyncio.Future:
         return self._result
 
+    def log(self, message: str) -> None:
+        self._logs.append(message)
+        try:
+            self._channel.send_no_reply(
+                "waitForEventInfo",
+                {
+                    "info": {
+                        "waitId": self._wait_id,
+                        "phase": "log",
+                        "message": message,
+                    },
+                },
+            )
+        except Exception:
+            pass
+
 
 def throw_on_timeout(timeout: float, exception: Exception) -> asyncio.Task:
     async def throw() -> None:
         await asyncio.sleep(timeout / 1000)
         raise exception
 
     return asyncio.create_task(throw())
+
+
+def format_log_recording(log: List[str]) -> str:
+    if not log:
+        return ""
+    header = " logs "
+    header_length = 60
+    left_length = math.floor((header_length - len(header)) / 2)
```

Comment:
we don't have math.floor upstream is it actually required here?

Reply:
Upstream we do `"o".repeat(2.9)` and String.prototype.repeat does floor internally.